File: frontend/src/lib/ml/features.ts
/**
 * The model's feature contract, in the browser.
 *
 * The TypeScript mirror of `backend/app/ml/features.py` and
 * `backend/app/ml/adapter.py`. Two devices carry a model trained on one corpus
 * onto pose data FormVision extracted itself, and both are reproduced here:
 *
 * 1. **Canonical orientation.** Every signal is restated so larger always means
 *    more of the fault. Four of the ten are negations of how the angle is
 *    normally quoted, and every one of those flips is load-bearing: get one
 *    wrong and the model still runs, still returns confident probabilities, and
 *    is silently backwards for that signal forever.
 *
 * 2. **Within-clip ranking.** Each value becomes its percentile within the
 *    clip's own distribution, which is unitless and so survives the fact that
 *    the training corpus quotes its angles on a different scale.
 *
 * **What differs from the backend, unavoidably.** The batch pipeline ranks each
 * frame against the whole clip. Live, only the past exists, so `RunningRanks`
 * ranks against the session so far. That asks the same question (is this
 * moment unusual compared with how you have been moving today) with less data
 * early on. It also means live and upload will not produce identical verdicts
 * for the same movement, and cannot.
 */

import type { RawFrameAngles } from "@/lib/analysis/angles";

/** Signals measured on both sides of the body. */
export const PAIRED_SIGNALS = [
  "knee_flexion",
  "hip_flexion",
  "ankle_openness",
  "valgus",
] as const;

/** Signals with a single value for the whole body. */
export const SINGLE_SIGNALS = ["torso_lean", "depth_phase"] as const;

export type PairedSignal = (typeof PAIRED_SIGNALS)[number];

/**
 * One frame's signals in canonical orientation. `null` means *not measured*,
 * never zero. A front-on clip has no ankle reading at all, and a zero there
 * would describe a perfectly neutral ankle rather than the absence of one.
 *
 * Larger means: knee/hip flexion, more closed; ankleOpenness, shin further over
 * the foot (heel lifting); valgus, knee further toward the midline; torsoLean,
 * further from upright; depthPhase, lower in the movement.
 */
export interface FrameSample {
  knee_flexion_left: number | null;
  knee_flexion_right: number | null;
  hip_flexion_left: number | null;
  hip_flexion_right: number | null;
  ankle_openness_left: number | null;
  ankle_openness_right: number | null;
  valgus_left: number | null;
  valgus_right: number | null;
  torso_lean: number | null;
  depth_phase: number | null;
}

const negate = (value: number | null): number | null =>
  value === null ? null : -value;

/**
 * Build the canonical sample for one frame of live analysis.
 *
 * A direct port of `adapter.py::frame_sample`. The four negations are the whole
 * point:
 *
 * - Knee and hip angles *fall* as the joint closes, so flexion is their
 *   negative. Left as-is, "more bent" would rank as "less".
 * - Hip height is measured upward from the ankles, so it falls as the lifter
 *   descends, whereas `depth_phase` must grow. The training corpus's own depth
 *   column runs the other way, which is exactly the trap: a rank transform
 *   carries an increasing reparameterisation through unchanged and inverts a
 *   decreasing one without complaint.
 *
 * Ankle openness and valgus need no flip: `angles.ts` already defines the ankle
 * angle as opening when the heel lifts, and knee lateral offset as positive
 * when the knee travels medially.
 */
export function frameSample(raw: RawFrameAngles): FrameSample {
  return {
    knee_flexion_left: negate(raw.leftKneeDeg),
    knee_flexion_right: negate(raw.rightKneeDeg),
    hip_flexion_left: negate(raw.leftHipDeg),
    hip_flexion_right: negate(raw.rightHipDeg),
    ankle_openness_left: raw.leftAnkleDeg,
    ankle_openness_right: raw.rightAnkleDeg,
    valgus_left: raw.leftKneeLateral,
    valgus_right: raw.rightKneeLateral,
    torso_lean: raw.torsoLeanDeg,
    depth_phase: negate(raw.hipHeight),
  };
}

/**
 * Average of a paired signal, tolerating one missing side.
 *
 * Side-on footage genuinely only tracks one leg, so requiring both would
 * discard most of a usable clip.
 */
export function meanOf(sample: FrameSample, signal: PairedSignal): number | null {
  const left = sample[`${signal}_left`];
  const right = sample[`${signal}_right`];
  const present = [left, right].filter((v): v is number => v !== null);
  if (present.length === 0) return null;
  return present.reduce((a, b) => a + b, 0) / present.length;
}

/**
 * Absolute left/right difference, or null if either side is missing.
 *
 * Unlike `meanOf` this cannot fall back to one side: a difference needs both
 * terms, and inventing the missing one would manufacture symmetry that was
 * never observed, the most misleading possible answer to an asymmetry question.
 */
export function gapOf(sample: FrameSample, signal: PairedSignal): number | null {
  const left = sample[`${signal}_left`];
  const right = sample[`${signal}_right`];
  if (left === null || right === null) return null;
  return Math.abs(left - right);
}

/** Flatten a sample into every rankable quantity, keyed as the model names them. */
export function derivedValues(sample: FrameSample): Record<string, number | null> {
  const values: Record<string, number | null> = {};

  for (const signal of PAIRED_SIGNALS) {
    values[`${signal}_left`] = sample[`${signal}_left`];
    values[`${signal}_right`] = sample[`${signal}_right`];
    values[`${signal}_mean`] = meanOf(sample, signal);
    values[`${signal}_gap`] = gapOf(sample, signal);
  }
  for (const signal of SINGLE_SIGNALS) {
    values[signal] = sample[signal];
  }

  return values;
}

/** Index of the first element not less than `value`. */
function lowerBound(sorted: number[], value: number): number {
  let low = 0;
  let high = sorted.length;
  while (low < high) {
    const mid = (low + high) >> 1;
    if (sorted[mid] < value) low = mid + 1;
    else high = mid;
  }
  return low;
}

/** Index of the first element greater than `value`. */
function upperBound(sorted: number[], value: number): number {
  let low = 0;
  let high = sorted.length;
  while (low < high) {
    const mid = (low + high) >> 1;
    if (sorted[mid] <= value) low = mid + 1;
    else high = mid;
  }
  return low;
}

/**
 * The session's own distribution of each quantity, grown one frame at a time.
 *
 * The live counterpart to the backend's `ClipReference`. Values are kept sorted
 * by binary-search insertion so a rank is a binary search rather than a scan,
 * which matters when ten quantities are ranked on every frame at 30fps.
 */
export class RunningRanks {
  private readonly distributions = new Map<string, number[]>();
  private count = 0;

  /** Frames observed so far, however few of their signals were measurable. */
  get frames(): number {
    return this.count;
  }

  /** Record one frame's values into the session distribution. */
  insert(values: Record<string, number | null>): void {
    this.count += 1;
    for (const [name, value] of Object.entries(values)) {
      if (value === null || Number.isNaN(value)) continue;
      let sorted = this.distributions.get(name);
      if (!sorted) {
        sorted = [];
        this.distributions.set(name, sorted);
      }
      sorted.splice(lowerBound(sorted, value), 0, value);
    }
  }

  /**
   * Where `value` falls in the session's distribution, as a 0-1 fraction, or
   * null if it is missing or the session never measured this quantity.
   *
   * Ties resolve to the **midpoint** of the tied run, and that detail is
   * load-bearing. Taking the upper edge would rank a signal that never varies at
   * 1.0 in every frame, i.e. maximally extreme, when the truth is the opposite:
   * a quantity that never moves has nothing remarkable about it anywhere. That
   * bug made a perfectly symmetric lifter register as asymmetric on every
   * repetition, because a left/right gap pinned at zero ranked at the top of its
   * own distribution.
   */
  rank(name: string, value: number | null): number | null {
    if (value === null || Number.isNaN(value)) return null;
    const sorted = this.distributions.get(name);
    if (!sorted || sorted.length === 0) return null;
    return (lowerBound(sorted, value) + upperBound(sorted, value)) / (2 * sorted.length);
  }

  reset(): void {
    this.distributions.clear();
    this.count = 0;
  }
}

/**
 * Build one detector's feature row from a frame.
 *
 * Feature names are read off the model rather than hardcoded, so retraining
 * with a different feature set needs no change here. Every shipped feature is a
 * `rank_` of a derived quantity; anything else is reported missing rather than
 * guessed at, which surfaces as low completeness and makes the detector abstain.
 *
 * Returns the row alongside the share of it that was actually measured.
 */
export function featureRow(
  featureNames: readonly string[],
  values: Record<string, number | null>,
  ranks: RunningRanks,
): { row: (number | null)[]; completeness: number } {
  const row = featureNames.map((name) => {
    if (!name.startsWith("rank_")) return null;
    return ranks.rank(name.slice("rank_".length), values[name.slice("rank_".length)]);
  });

  if (row.length === 0) return { row, completeness: 0 };
  const present = row.filter((value) => value !== null).length;
  return { row, completeness: present / row.length };
}

File: backend/app/ml/features.py
"""The feature vector shared by training and inference.

This module is the seam that makes the ML layer trustworthy, and it is the one
place in the ML code worth reading carefully. It is imported by *both* the
offline training scripts and the live pipeline, so there is exactly one
definition of what a feature means. Train/serve skew cannot be introduced by
editing one side of the system, because there is only one side.

Like `analysis/types.py`, nothing here imports scikit-learn, numpy, or anything
else heavyweight. A `FrameSample` is plain floats, so the feature builder can be
tested by hand.

**The problem this module solves.** The training corpus was extracted by a
different tool, from videos we do not have, using angle conventions that do not
match FormVision's: its knee angle has a median of 54 degrees and a floor of 5,
where FormVision's is ~175 standing and ~80 at depth. The two are not the same
function of the same joint, so a model fitted on raw values from one would be
meaningless applied to the other.

Two devices bridge that gap:

1. **Canonical orientation.** Every signal is restated so that *larger always
   means more of the thing the fault is about* (see `FrameSample`). Each adapter
   is responsible for flipping its own source's sign. This has to be explicit:
   the training corpus's depth signal grows as the lifter descends while
   FormVision's shrinks, and a rank transform preserves an increasing
   reparameterisation but silently inverts a decreasing one.

2. **Within-clip ranking.** Each value is expressed as its percentile within the
   clip's own distribution of that signal. A rank is invariant to *any* monotone
   reparameterisation, where a z-score would only survive an affine one, which is
   why ranking is the correct tool here rather than merely the convenient one.

Absolute features are kept alongside the ranks, because ranking has a real
weakness: a lifter whose knees cave on every single rep has a flat within-clip
distribution, and the fault disappears into it. Which family actually earns its
place is decided per fault by cross-validation, not by argument here.
"""

from __future__ import annotations

from bisect import bisect_left, bisect_right
from collections.abc import Iterable, Mapping, Sequence
from dataclasses import dataclass, fields

#: Signals measured on both sides of the body.
PAIRED_SIGNALS: tuple[str, ...] = (
    "knee_flexion",
    "hip_flexion",
    "ankle_openness",
    "valgus",
)

#: Signals with a single value for the whole body.
SINGLE_SIGNALS: tuple[str, ...] = (
    "torso_lean",
    "depth_phase",
)


@dataclass(frozen=True, slots=True)
class FrameSample:
    """One frame's signals, in canonical orientation.

    Every field is `None`-able and `None` means *not measured*, never zero. A
    front-on clip has no ankle reading at all, and a zero there would describe a
    perfectly neutral ankle rather than the absence of one.

    Canonical orientation, which every adapter must honour:

    ==================  ==========================================
    Field               Larger means
    ==================  ==========================================
    knee_flexion        knee more bent (deeper)
    hip_flexion         hip more closed (more folded)
    ankle_openness      shin further over the foot; heel lifting
    valgus              knee further inward, toward the midline
    torso_lean          torso further from upright
    depth_phase         lifter lower in the movement
    ==================  ==========================================

    Note that four of these are negations of how the underlying angle is
    normally quoted. A knee angle *falls* as the knee bends, so `knee_flexion`
    is its negative. Adapters do that flip; this module never guesses.
    """

    knee_flexion_left: float | None = None
    knee_flexion_right: float | None = None
    hip_flexion_left: float | None = None
    hip_flexion_right: float | None = None
    ankle_openness_left: float | None = None
    ankle_openness_right: float | None = None
    valgus_left: float | None = None
    valgus_right: float | None = None
    torso_lean: float | None = None
    depth_phase: float | None = None

    def value(self, name: str) -> float | None:
        """Look a signal up by name, for the generic feature loops below."""
        return getattr(self, name)

    def paired(self, signal: str) -> tuple[float | None, float | None]:
        """The left and right readings of a paired signal."""
        return (getattr(self, f"{signal}_left"), getattr(self, f"{signal}_right"))

    def mean_of(self, signal: str) -> float | None:
        """Average of a paired signal, tolerating one missing side.

        Side-on footage genuinely only tracks one leg, so requiring both would
        discard most of a usable clip. See `CORE_LANDMARKS` in `analysis/types`
        for the same reasoning applied to landmarks.
        """
        present = [value for value in self.paired(signal) if value is not None]
        if not present:
            return None
        return sum(present) / len(present)

    def gap_of(self, signal: str) -> float | None:
        """Absolute left/right difference, or None if either side is missing.

        Unlike `mean_of`, this cannot fall back to one side: a difference needs
        both terms, and inventing the missing one would manufacture symmetry
        that was never observed.
        """
        left, right = self.paired(signal)
        if left is None or right is None:
            return None
        return abs(left - right)


#: Derived quantities computed from a `FrameSample` before ranking. Named here
#: so `ClipReference` and the feature builder agree on the vocabulary without
#: either having to know the other's internals.
def derived_values(sample: FrameSample) -> dict[str, float | None]:
    """Flatten a sample into the full set of rankable quantities."""
    values: dict[str, float | None] = {}

    for signal in PAIRED_SIGNALS:
        left, right = sample.paired(signal)
        values[f"{signal}_left"] = left
        values[f"{signal}_right"] = right
        values[f"{signal}_mean"] = sample.mean_of(signal)
        values[f"{signal}_gap"] = sample.gap_of(signal)

    for signal in SINGLE_SIGNALS:
        values[signal] = sample.value(signal)

    return values


#: Every quantity that gets both an absolute and a ranked feature, in a fixed
#: order. Feature vectors are positional once they reach scikit-learn, so this
#: ordering is part of the model contract and must not be reshuffled without
#: retraining.
QUANTITY_NAMES: tuple[str, ...] = tuple(
    [
        f"{signal}_{suffix}"
        for signal in PAIRED_SIGNALS
        for suffix in ("left", "right", "mean", "gap")
    ]
    + list(SINGLE_SIGNALS)
)


@dataclass(frozen=True, slots=True)
class ClipReference:
    """The clip's own distribution of each quantity, for ranking against.

    Built once per clip and reused for every frame in it. Values are stored
    sorted so a rank is a binary search rather than a scan, which matters when a
    two-minute clip at 30fps asks for eighteen ranks on each of 3,600 frames.
    """

    distributions: Mapping[str, tuple[float, ...]]

    @classmethod
    def from_samples(cls, samples: Iterable[FrameSample]) -> ClipReference:
        """Collect each quantity's observed values across a clip.

        Missing readings are skipped rather than filled. A quantity that was
        never measurable ends up with an empty distribution, and `rank` then
        returns None for it, which is what carries "this camera angle could not
        see it" all the way through to the predictor abstaining.
        """
        collected: dict[str, list[float]] = {name: [] for name in QUANTITY_NAMES}

        for sample in samples:
            for name, value in derived_values(sample).items():
                if value is not None:
                    collected[name].append(value)

        return cls(
            distributions={
                name: tuple(sorted(values)) for name, values in collected.items()
            }
        )

    def rank(self, name: str, value: float | None) -> float | None:
        """Where `value` falls in the clip's distribution, as a 0-1 fraction.

        Ties resolve to the *midpoint* of the tied run rather than its end, and
        that detail is load-bearing. Taking the upper edge would rank a signal
        that never varies at 1.0 in every frame, i.e. maximally extreme, when
        the truth is the exact opposite: a quantity that never moves has nothing
        remarkable about it anywhere. That bug made a perfectly symmetric lifter
        register as asymmetric on every repetition, because a left/right gap
        pinned at zero ranked at the top of its own distribution.

        With the midpoint convention a constant signal ranks 0.5 throughout,
        which reads as "unremarkable" and is what the model was trained to
        expect from ordinary movement.

        Returns None when the value is missing or the clip never measured this
        quantity at all.
        """
        if value is None:
            return None

        ordered = self.distributions.get(name)
        if not ordered:
            return None

        low = bisect_left(ordered, value)
        high = bisect_right(ordered, value)
        return (low + high) / (2 * len(ordered))


#: The full ordered feature vector. Absolute features first, then ranks, so a
#: model card listing the names reads in a predictable order.
FEATURE_NAMES: tuple[str, ...] = tuple(
    [f"abs_{name}" for name in QUANTITY_NAMES]
    + [f"rank_{name}" for name in QUANTITY_NAMES]
)


def build_frame_features(
    sample: FrameSample, reference: ClipReference
) -> dict[str, float | None]:
    """Build one frame's complete feature vector.

    The single function called by both the training pipeline and the live
    predictor. Keys are exactly `FEATURE_NAMES`; values are `None` wherever the
    underlying signal was not measurable.
    """
    values = derived_values(sample)

    features: dict[str, float | None] = {}
    for name in QUANTITY_NAMES:
        features[f"abs_{name}"] = values[name]
        features[f"rank_{name}"] = reference.rank(name, values[name])

    return features


def feature_row(
    features: Mapping[str, float | None], names: Sequence[str]
) -> tuple[list[float | None], float]:
    """Project a feature mapping onto an ordered subset, with a completeness score.

    Returns the values in `names` order alongside the fraction of them that were
    actually measured. The predictor uses that fraction to decide whether it
    knows enough to speak: a valgus verdict resting on two of six features is
    not a verdict, and the honest output is silence.
    """
    row = [features.get(name) for name in names]
    if not row:
        return row, 0.0

    present = sum(1 for value in row if value is not None)
    return row, present / len(row)


#: Which features each fault detector is allowed to see.
#:
#: **Ranks only, deliberately.** The absolute features above are computed and
#: available, but none of them appears here, because an absolute value cannot
#: cross the gap between the training corpus and FormVision. The corpus quotes
#: its knee lateral offset in raw normalised-coordinate units; FormVision
#: divides its own by torso length. A decision boundary learned at 0.05 in one
#: set of units is not the same boundary in the other, and nothing at runtime
#: would reveal the mismatch. Ranks are unitless by construction, so they are
#: the only family that genuinely transfers.
#:
#: The cost is a real and documented blind spot: a lifter whose knees cave on
#: *every* rep has a flat within-clip distribution, so their valgus ranks
#: mid-range and the detector stays quiet. Ranking finds the fault that stands
#: out from a lifter's own movement, not the one baked into all of it. Catching
#: the persistent case needs an absolute threshold in FormVision's own units,
#: which is a rule, not a model.
#:
#: Restricting inputs per fault is also what keeps each detector interpretable
#: and makes a fault the camera cannot see fail cleanly on completeness rather
#: than quietly substituting a correlated signal.
#:
#: **Each shipped detector sees its own signal and nothing else that moves with
#: the label.** This is not tidiness, it is the result of an ablation. The
#: corpus perturbs knee *angle* at the same time as knee lateral offset when it
#: manufactures a caving-knees example, so a valgus detector given the mean knee
#: flexion reached 0.994 AUC while scoring 0.910 from that knee flexion alone.
#: It was mostly detecting the corpus's bookkeeping. Real caving knees do not
#: come with extra knee bend attached, so that detector would have collapsed on
#: real footage while looking excellent in cross-validation. The same held for
#: asymmetry, where perturbing each side independently shifts the mean as well
#: as the gap. Both mean terms are gone.
#:
#: The depth-phase rank stays in all of them, and is the one shared input.
#: The corpus is per-frame with no notion of where in a rep a frame sits, so
#: without it a correct standing frame and a faulty bottom frame are the same
#: point. It is safe to share precisely because it carries no label signal: the
#: corpus never recomputes its depth column after perturbing anything, so it is
#: bit-identical across all six labels of a frame and scores exactly 0.500 AUC
#: on its own against every fault. It contributes context, never evidence.
FAULT_FEATURES: Mapping[str, tuple[str, ...]] = {
    "knee_valgus": (
        "rank_valgus_left",
        "rank_valgus_right",
        "rank_valgus_mean",
        "rank_valgus_gap",
        "rank_depth_phase",
    ),
    "heel_lift": (
        "rank_ankle_openness_left",
        "rank_ankle_openness_right",
        "rank_ankle_openness_mean",
        "rank_depth_phase",
    ),
    "asymmetry": (
        "rank_knee_flexion_gap",
        "rank_hip_flexion_gap",
        "rank_depth_phase",
    ),
    # Trained and reported for the model card, but not wired to feedback:
    # FormVision measures depth and lean directly and correctly, and a direct
    # measurement beats a model fitted on synthetic offsets.
    "shallow": (
        "rank_knee_flexion_mean",
        "rank_hip_flexion_mean",
        "rank_depth_phase",
    ),
    "forward_lean": (
        "rank_torso_lean",
        "rank_depth_phase",
    ),
}

#: The faults whose predictions become coaching feedback. The rest are trained
#: for evaluation only.
SHIPPED_FAULTS: tuple[str, ...] = ("knee_valgus", "heel_lift", "asymmetry")


def sample_field_names() -> tuple[str, ...]:
    """Field names of `FrameSample`, for adapters that build one generically."""
    return tuple(field.name for field in fields(FrameSample))
